### PR TITLE
[JENKINS-39355 follow up] Adapting to the SCM and Branch 2.0.x APIs

### PR DESCRIPTION
**Pull request by stephenc**

### CODE REVIEW

**Comment by stephenc** [comment]
@reviewbybees

**Comment by rsandell** [comment]
quick :bee:

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java [hunk 1]
```diff
@@ -66,11 +67,17 @@ synchronized void setBranch(@Nonnull Branch branch) {
             @Override public boolean hasPermission(Authentication a, Permission permission) {
                 // This project is managed by its parent and may not be directly configured or deleted.
                 // Note that Item.EXTENDED_READ may still be granted, so you can still see Snippet Generator, etc.
-                if (ACL.SYSTEM.equals(a)) {
+                if (branch instanceof Branch.Dead && (permission == Job.BUILD)) {
+                    // if dead, nobody, not even SYSTEM may build this job
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java]: Is this not better covered by `isBuildable`?

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java [hunk 2]
```diff
@@ -66,11 +67,17 @@ synchronized void setBranch(@Nonnull Branch branch) {
...
+                    return false;
+                } else if (ACL.SYSTEM.equals(a)) {
                     return true; // e.g., DefaultDeadBranchStrategy.runDeadBranchCleanup
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java]: Is this not now covered by the check below? I guess it is safest to unconditionally allow `SYSTEM` to do what it needs, but would be advisable to update comments to reflect the new reality.

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java [hunk 3]
```diff
@@ -66,11 +67,17 @@ synchronized void setBranch(@Nonnull Branch branch) {
...
-                } else if (permission == Item.CONFIGURE || permission == Item.DELETE) {
+                } else if (permission == Item.CONFIGURE) {
                     return false;
-                } else {
+                } else if (permission == Item.DELETE) {
+                    // allow early manual clean-up of dead branches
+                    return branch instanceof Branch.Dead && acl.hasPermission(a, permission);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java]: Perhaps more intuitively:

```java
} else if (permission == Item.DELETE && !(branch instanceof Branch.Dead)) {
    return false;
```

else we fall through to the normal ACL check.

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowBranchProjectFactory.java [hunk 4]
```diff
@@ -48,7 +49,33 @@
...
+                        if (stat.getAlternativePath() != null) {
+                            listener.getLogger().format("      ‘%s’ not found (but found ‘%s’, search is case sensitive)%n", SCRIPT, stat.getAlternativePath());
+                        } else {
+                            listener.getLogger().format("      ‘%s’ not found%n",SCRIPT);
+                        }
+                        return false;
+                    case DIRECTORY:
+                        listener.getLogger().format("      ‘%s’ found but is a directory not a file%n", SCRIPT);
+                        return false;
+                    default:
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowBranchProjectFactory.java]: Should this not be checking `FILE` only?

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java]: and how exactly do you get there from here?

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowBranchProjectFactory.java (hunk 4 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowBranchProjectFactory.java]: It is checking `REGULAR_FILE`, `LINK` and `OTHER` as these should be file-like enough

**Comment by rsandell** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/BranchJobProperty.java]: cf. #53

**Pull request merged**